### PR TITLE
Reengineer Nailgun

### PR DESCRIPTION
**Pull request by sbalabanov**
Restructured file structure, format google codestyle, added tests (JUnit5) and mocks (Mockito), refactored sessions, bug fixes

### CODE REVIEW

File: nailgun-examples/src/main/java/com/martiansoftware/nailgun/examples/Hash.java [hunk 1]
```diff
@@ -18,93 +18,93 @@
...
+   * Provides a list of algorithms for the specified service (which, for our purposes, is
+   * "MessageDigest".
+   *
+   * <p>This method was only very slightly adapted (to use a TreeSet) from the Java Almanac at
+   * http://javaalmanac.com/egs/java.security/ListServices.html
+   *
+   * @param serviceType The name of the service we're looking for. It's "MessageDigest"
+   */
+  private static Set getCryptoImpls(String serviceType) {
+    Set result = new java.util.TreeSet();
```
Comment (ttsugriy) [inline nailgun-examples/src/main/java/com/martiansoftware/nailgun/examples/Hash.java]: nit: why fully qualified import? Does it conflict with something?

File: nailgun-examples/src/main/java/com/martiansoftware/nailgun/examples/Hash.java [hunk 2]
```diff
@@ -18,93 +18,93 @@
...
+
+    // All all providers
+    Provider[] providers = Security.getProviders();
+    for (int i = 0; i < providers.length; i++) {
+      // Get services provided by each provider
+      Set keys = providers[i].keySet();
+      for (Iterator it = keys.iterator(); it.hasNext(); ) {
```
Comment (ttsugriy) [inline nailgun-examples/src/main/java/com/martiansoftware/nailgun/examples/Hash.java]: nit: why not `foreach`?
```
for (String key : providers[i].keySet()
 ...
```

File: nailgun-examples/src/main/java/com/martiansoftware/nailgun/examples/Hash.java [hunk 3]
```diff
@@ -18,93 +18,93 @@
...
-    		StringBuffer buf = new StringBuffer();
-    		for (int i = 0; i < result.length; ++i) {
-    			buf.append(HEXCHARS[(result[i] >> 4) & 0x0f]);
-    			buf.append(HEXCHARS[result[i] & 0x0f]);
-    		}
-    		context.out.println(buf);
-    	}
+      }
     }
+    return (result);
```
Comment (ttsugriy) [inline nailgun-examples/src/main/java/com/martiansoftware/nailgun/examples/Hash.java]: what's up with the `()`?

File: nailgun-examples/src/main/java/com/martiansoftware/nailgun/examples/Hash.java (hunk 1 above)
Comment (ttsugriy) [reply nailgun-examples/src/main/java/com/martiansoftware/nailgun/examples/Hash.java]: also, why is it not generic? :)

File: nailgun-examples/src/main/java/com/martiansoftware/nailgun/examples/DumpAll.java [hunk 4]
```diff
@@ -18,37 +18,36 @@
...
+    context.out.println("   context.getFileSeparator(): " + context.getFileSeparator());
+    context.out.println("   context.getPathSeparator(): " + context.getPathSeparator());
+
+    context.out.println("\ncontext.getArgs():");
+    for (int i = 0; i < context.getArgs().length; ++i) {
+      context.out.println("   args[" + i + "]=" + context.getArgs()[i]);
+    }
+
+    context.out.println("\ncontext.getEnv():");
+    TreeSet keys = new TreeSet(context.getEnv().keySet());
```
Comment (ttsugriy) [inline nailgun-examples/src/main/java/com/martiansoftware/nailgun/examples/DumpAll.java]: generics

File: nailgun-examples/src/main/java/com/martiansoftware/nailgun/examples/DumpAll.java [hunk 5]
```diff
@@ -18,37 +18,36 @@
...
+    for (Iterator i = keys.iterator(); i.hasNext(); ) {
```
Comment (ttsugriy) [inline nailgun-examples/src/main/java/com/martiansoftware/nailgun/examples/DumpAll.java]: foreach

File: nailgun-server/src/main/java/com/martiansoftware/nailgun/NGListeningAddress.java [hunk 6]
```diff
@@ -99,4 +100,16 @@ public String toString() {
...
     }
   }
+
+  /**
+   * Close any instances of local socket, i.e. Unix socket or Windows named pipe
+   *
+   * @param localAddress
+   */
+  public static void release(String localAddress) {
+    if (Platform.isWindows()) {
```
Comment (ttsugriy) [inline nailgun-server/src/main/java/com/martiansoftware/nailgun/NGListeningAddress.java]: please add a comment saying that for windows there is nothing to do

File: nailgun-server/pom.xml [hunk 7]
```diff
@@ -41,6 +41,12 @@
         <version>5.2.0</version>
         <scope>test</scope>
       </dependency>
+      <dependency>
+        <groupId>org.mockito</groupId>
```
Comment (ttsugriy) [inline nailgun-server/pom.xml]: prefer fakes over mocks ;)

File: nailgun-examples/src/main/java/com/martiansoftware/nailgun/examples/Hash.java (hunk 1 above)
Comment (sbalabanov) [reply nailgun-examples/src/main/java/com/martiansoftware/nailgun/examples/Hash.java]: Only code formatting changes were in 'examples' folder

**Comment by ttsugriy** [comment]
looks like appveyor tests are failing - https://ci.appveyor.com/project/Facebook/nailgun/build/95#L1199 but otherwise looks legit

**Pull request merged**